### PR TITLE
Adding reverse and symmetric KLD losses

### PR DESCRIPTION
**Pull request by insop**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* Adding reverse and symmetric KLD loss
* Adding KLD losses based on [link](https://github.com/jongwooko/distillm/blob/17c0f98bc263b1861a02d5df578c84aea652ee65/distillm/losses.py)

#### Test plan

| Model | #Params | Method | HumanEval base | HumanEval plus | MBPP base | MBPP plus |
|:--|--:|:--|--:|--:|--:|--:|
| Llama3.2 | 8B | Teacher (Llama3.1) | 39.0 | 34.1 | 64.3 | 53.2 |
| | 1B | Base model | 18.3 | 15.9 | 35.4 | 29.0 |
| | | FT w/o KD | 22.0 | 17.7 | 37.6 | 31.7 |
| | | Forward KLD | 22.6 | 19.5 | **41.3** | 33.1 |
| | | Reverse KLD | 20.7 | 17.7 | 39.2 | 33.6 |
| | | Symmetric KLD | **23.2** | **20.1** | 41.0 | **33.9** |

Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [x] I did not change any public API
- [x] I have added an example to docs or docstrings

### CODE REVIEW

**Comment by insop** [comment]
@ebsmothers , PTAL.
Thank you.

File: torchtune/modules/loss/kd_losses.py [hunk 1]
```diff
@@ -58,6 +58,105 @@ def forward(
...
+            return torch.tensor(0.0, device=x.device)
+        return -torch.sum(x * mask.view(-1), dim=0) / torch.sum(mask.view(-1), dim=0)
+
+class SymmetricKLLoss(torch.nn.Module):
+    """
+    The Symmetric Kullback-Leibler divergence loss for valid indexes.
+    Implementation of https://github.com/jongwooko/distillm/blob/17c0f98bc263b1861a02d5df578c84aea652ee65/distillm/losses.py
+
+    Args:
+        sym_kd_ratio (float): Ratio of symmetric KL divergence loss.
```
Comment (ebsmothers) [inline torchtune/modules/loss/kd_losses.py:121]: nit: let's be a bit more explicit, e.g. "When set to 1 this loss reduces to forward KL divergence, when set to 0 this loss reduces to reverse kl divergence". Also separately it'd be good to do a value check that `0 <= sym_kd_ratio <= 1` on init.

File: torchtune/modules/loss/kd_losses.py [hunk 2]
```diff
@@ -138,3 +237,164 @@ def forward(
             )
 
         return total_fkl_loss / torch.sum(mask.view(-1), dim=0)
+
+class ReverseKLWithChunkedOutputLoss(torch.nn.Module):
```
Comment (ebsmothers) [inline torchtune/modules/loss/kd_losses.py:241]: Not necessary for this PR but as we are starting to have a proliferation of chunked loss implementations I wonder whether it'd be worth investing in a general utility to wrap an arbitrary loss with chunking operation @felipemello1

File: torchtune/modules/loss/kd_losses.py [hunk 3]
```diff
@@ -138,3 +237,164 @@ def forward(
...
+
+        total_rkl_loss = 0.0
+        for student_chunk, teacher_chunk, label_chunk in zip(
+            student_logits, teacher_logits, labels
+        ):
+            total_rkl_loss += self.rkl_loss(
+                student_chunk, teacher_chunk, label_chunk, normalize=False
+            )
+
+        return total_rkl_loss / torch.sum(mask.view(-1), dim=0)
```
Comment (ebsmothers) [inline torchtune/modules/loss/kd_losses.py:319]: I don't think our existing chunked forward KL loss even does this, but I wonder why we don't do the same check that `torch.sum(mask.view(-1), dim=0) != 0` that we're doing in the unchunked version? Couldn't we still potentially get division by zero here?

File: torchtune/training/_compile.py [hunk 4]
```diff
@@ -77,6 +79,10 @@ def compile_loss(loss: nn.Module, verbose: bool = True) -> nn.Module:
         )
     elif isinstance(loss, ForwardKLWithChunkedOutputLoss):
         loss.fkl_loss = torch.compile(loss.fkl_loss, backend=backend)
+    elif isinstance(loss, ReverseKLWithChunkedOutputLoss):
+        loss.rkl_loss = torch.compile(loss.rkl_loss, backend=backend)
+    elif isinstance(loss, SymmetricKLWithChunkedOutputLoss):
+        loss.sym_kl_loss = torch.compile(loss.sym_kl_loss, backend=backend)
```
Comment (ebsmothers) [inline torchtune/training/_compile.py:85]: Kinda related to my comment above, we should also consider just defining a protocol `ChunkedLoss` or something having a method e.g. `def base_loss(self, *args, **kwargs)` and having all of `CEWithChunkedOutputLoss`, `ForwardKLWithChunkedOutputLoss`, `ReverseKLWithChunkedOutputLoss`, `SymmetricKLWithChunkedOutputLoss` inherit from that. Then L76-L85 basically just become 

```
if isinstance(loss, ChunkedLoss):
	loss.base_loss = torch.compile(loss.base_loss, backend=backend)
```

Again, not a blocker for this particular PR.

File: tests/torchtune/modules/loss/test_kd_losses.py [hunk 5]
```diff
@@ -114,3 +114,201 @@ def test_forward_kl_loss_expected(self):
         # assert
         assert_expected(chunked_loss, expected_loss, rtol=1e-2, atol=1e-2)
         assert_expected(standard_loss, expected_loss, rtol=1e-2, atol=1e-2)
+
+class TestReverseKLWithChunkedOutputLoss:
```
Comment (ebsmothers) [inline tests/torchtune/modules/loss/test_kd_losses.py:118]: Thanks for adding these unit tests!

File: tests/torchtune/modules/loss/test_kd_losses.py [hunk 6]
```diff
@@ -114,3 +114,201 @@ def test_forward_kl_loss_expected(self):
...
+                    [1.5156, 0.1641, 2.0781, -0.7734],
+                    [-0.5898, 0.4453, -0.7969, 0.6328],
+                    [0.6289, -0.8359, 0.9258, 0.2109],
+                    [0.0006, 0.5195, 3.2344, -1.5781],
+                ],
+            ],
+            dtype=torch.bfloat16,
+        )
+        labels = torch.tensor([[0, 3, 3, 1], [1, 1, 1, 1]])
+        expected_loss = torch.tensor(0.6775, dtype=torch.float32)
```
Comment (ebsmothers) [inline tests/torchtune/modules/loss/test_kd_losses.py:193]: Just to verify: did you set this value based on the reference implementation in distillm?

**Review by ebsmothers — COMMENTED** [review]
Thanks @insop for the PR! I left a few comments but no major concerns. One thing you'll need to fix is the failing linter job -- if you haven't already you can set up and run pre-commit on all your modified files by following [this section](https://github.com/pytorch/torchtune/blob/32e265d5749fd592711a03247486eafa6c898d94/CONTRIBUTING.md#coding-style) of our contributing guide (assuming you already performed a dev install). If you have any trouble do let me know and we can help out.

**Comment by insop** [comment]
Thank you for the review and comments, @ebsmothers.
Ack and will soon follow up the comments.

File: torchtune/modules/loss/kd_losses.py (hunk 3 above)
Comment (lindawangg) [reply torchtune/modules/loss/kd_losses.py:319]: We didn't add this check for the chunked forward KL either, mainly because if all the labels are ignore index, there's probably something wrong with the training data. Wondering if we should add this check in the training module to cover all the loss cases, maybe in a separate diff?

**Comment by lindawangg** [comment]
@insop do you have any results training with the losses that you could add to the test plan?

File: torchtune/modules/loss/kd_losses.py (hunk 1 above)
Comment (insop) [reply torchtune/modules/loss/kd_losses.py:121]: My apologies for the long delay.

I will update the code accordingly and push soon.

```
+            When set to 1 this loss reduces to forward KL divergence, when set to 0 this loss reduces to reverse kl divergence.
+
         ignore_index (int):  Specifies a target value that is ignored and does not contribute to the input gradient.
             The loss is divided over non-ignored targets.
             Default: -100.
@@ -126,6 +128,8 @@ class SymmetricKLLoss(torch.nn.Module):
 
     def __init__(self, sym_kd_ratio: float = 0.5, ignore_index: int = -100):
         super().__init__()
+        if sym_kd_ratio < 0.0 or sym_kd_ratio > 1.0:
+            raise ValueError("sym_kd_ratio must be in the range [0, 1]")
```

File: torchtune/modules/loss/kd_losses.py (hunk 3 above)
Comment (insop) [reply torchtune/modules/loss/kd_losses.py:319]: Created an issue for tracking, https://github.com/pytorch/torchtune/issues/2225

File: torchtune/training/_compile.py (hunk 4 above)
Comment (insop) [reply torchtune/training/_compile.py:85]: Created an issue for tracking, https://github.com/pytorch/torchtune/issues/2226

File: tests/torchtune/modules/loss/test_kd_losses.py (hunk 6 above)
Comment (insop) [reply tests/torchtune/modules/loss/test_kd_losses.py:193]: Yes, I've tested against [distillm](https://github.com/jongwooko/distillm).
My initial test code is [here](https://gist.github.com/insop/aa565f463cd4fbfcd8f6e2d754246130)

**Comment by insop** [comment]
> Thanks @insop for the PR! I left a few comments but no major concerns. One thing you'll need to fix is the failing linter job -- if you haven't already you can set up and run pre-commit on all your modified files by following [this section](https://github.com/pytorch/torchtune/blob/32e265d5749fd592711a03247486eafa6c898d94/CONTRIBUTING.md#coding-style) of our contributing guide (assuming you already performed a dev install). If you have any trouble do let me know and we can help out.

My apologies for the long delay. Addresses pre-commit check and review comments.

**Comment by insop** [comment]
> @insop do you have any results training with the losses that you could add to the test plan?

@lindawangg 

My apologies for the long delay. Here is one example of running a small-scale test using Llama 3.1 8B as a teacher and Llama 3.2 1B as a student with the Code Alpaca dataset.

I'm not entirely sure about 'adding to the test plan' in this case. Could you please clarify your suggestion?


| Model | #Params | Method | HumanEval base | HumanEval plus | MBPP base | MBPP plus |
|:--|--:|:--|--:|--:|--:|--:|
| Llama3.2 | 8B | Teacher (Llama3.1) | 39.0 | 34.1 | 64.3 | 53.2 |
| | 1B | Base model | 18.3 | 15.9 | 35.4 | 29.0 |
| | | FT w/o KD | 22.0 | 17.7 | 37.6 | 31.7 |
| | | Forward KLD | 22.6 | 19.5 | **41.3** | 33.1 |
| | | Reverse KLD | 20.7 | 17.7 | 39.2 | 33.6 |
| | | Symmetric KLD | **23.2** | **20.1** | 41.0 | **33.9** |

**Comment by insop** [comment]
Let me know if you have any comments.

**Comment by ebsmothers** [comment]
@insop sorry for the delayed response. Re "adding to the test plan", I think it's just referring to updating the PR summary to show your results. Let me quickly update it for you based on your previous comment. Otherwise it appears that your linter job is still failing, lmk if you need any help with this.

**Comment by insop** [comment]
Thank you @ebsmothers, I have updated the lint issue.

I used default config files and argument overriding for my training. I will put them together for others to use shortly.
I will update the PR based on your comment about my config in a day or two.

**Comment by insop** [comment]
Hi @ebsmothers 

Here is populated config file that can be used as is

<details><summary>config file</summary>

```yaml
batch_size: 4
checkpointer:
  _component_: torchtune.training.FullModelHFCheckpointer
  checkpoint_dir: ...
  checkpoint_files:
  - model.safetensors
  model_type: LLAMA3_2
  output_dir: ...
  recipe_checkpoint: null
  safe_serialization: true
compile: false
dataset:
  _component_: torchtune.datasets.alpaca_code_instruction_120k_dataset
  packed: false
device: cuda
dtype: bf16
enable_activation_checkpointing: true
epochs: 1
gradient_accumulation_steps: 8
kd_loss:
  _component_: torchtune.modules.loss.ForwardKLWithChunkedOutputLoss
kd_ratio: 0.75
log_every_n_steps: 1
log_peak_memory_stats: true
loss:
  _component_: torchtune.modules.loss.CEWithChunkedOutputLoss
lr_scheduler:
  _component_: torchtune.training.lr_schedulers.get_cosine_schedule_with_warmup
  num_warmup_steps: 100
max_steps_per_epoch: null
metric_logger:
  _component_: torchtune.training.metric_logging.WandBLogger
  project: kd-llm
model:
  _component_: torchtune.models.llama3_2.lora_llama3_2_1b
  apply_lora_to_mlp: true
  apply_lora_to_output: false
  lora_alpha: 128
  lora_attn_modules:
  - q_proj
  - v_proj
  - output_proj
  lora_dropout: 0.0
  lora_rank: 64
optimizer:
  _component_: torch.optim.AdamW
  fused: true
  lr: 0.0003
  weight_decay: 0.01
output_dir: ...
profiler:
  _component_: torchtune.training.setup_torch_profiler
  active_steps: 2
  cpu: true
  cuda: true
  enabled: false
  num_cycles: 1
  output_dir: ...
  profile_memory: false
  record_shapes: true
  wait_steps: 5
  warmup_steps: 3
  with_flops: false
  with_stack: false
resume_from_checkpoint: false
save_adapter_weights_only: false
seed: 1234
shuffle: true
teacher_checkpointer:
  _component_: torchtune.training.FullModelHFCheckpointer
  checkpoint_dir: ...
  checkpoint_files:
  - model-00001-of-00004_0.safetensors
  - model-00002-of-00004_0.safetensors
  - model-00003-of-00004_0.safetensors
  - model-00004-of-00004_0.safetensors
  model_type: LLAMA3
  output_dir: ...
  recipe_checkpoint: null
  safe_serialization: true
teacher_model:
  _component_: torchtune.models.llama3_1.llama3_1_8b
tokenizer:
  _component_: torchtune.models.llama3.llama3_tokenizer
  max_seq_len: null
  path: ...
```

</details> 

Here is helper script that used for update the config file during the experiments:


<details><summary>Helper script</summary>

```bash
#!/usr/bin/bash

set -e
set -u
set -o pipefail

# =========
# TORCHTUNE
# update
# =========
TRAIN_METHOD=knowledge_distillation_distributed
CONFIG_FILE="kd/kd_llama3_2_config.yaml"

# =========
## EXPERIMENT PARAMETERS
# update
# =========
# KD CONFIG
KD_RATIO=0.75

# =========
## MODEL
# update
# =========
# STUDENT MODEL
CHECKPOINT_DIR="..."
# TEACHER MODEL
TEACHER_CHECKPOINT_DIR="../8b_lora-Llama-3.1-8B-alpaca-code-120k"

# =========
# take the file name and use as a train name
TRAIN_NAME="${BASH_SOURCE[0]##*/}"
TRAIN_NAME="${TRAIN_NAME%.*}"

# Env setup
WORK_DIR=".."
CONFIG_DIR="$WORK_DIR/configs/llama3_2"
OUTPUT_BASE_DIR="/workspace/outputs_kd-llm/"
TOKENIZER_DIR="$OUTPUT_BASE_DIR/tokens_llama-3.2/"
TIMESTAMP=$(date +%Y%m%d-%H%M)

# dist
NNMODES=1
NPROC_PER_NODE=7
DIST_TRAIN_OPTIONS="--nnodes $NNMODES --nproc_per_node $NPROC_PER_NODE"

# =========

MODEL_NAME=$(basename "$CHECKPOINT_DIR")
CHECKPOINT_OUTPUT_DIR="$OUTPUT_BASE_DIR/${TRAIN_NAME}-$TIMESTAMP"

TEACHER_MODEL=$(basename "$TEACHER_CHECKPOINT_DIR")

CONFIG_FILE_OPTION="--config $CONFIG_DIR/$CONFIG_FILE"

CONFIG_OPTIONS=""
CONFIG_OPTIONS+=" checkpointer.checkpoint_dir=$CHECKPOINT_DIR"
CONFIG_OPTIONS+=" checkpointer.output_dir=$CHECKPOINT_OUTPUT_DIR"
CONFIG_OPTIONS+=" teacher_checkpointer.checkpoint_dir=$TEACHER_CHECKPOINT_DIR"
CONFIG_OPTIONS+=" kd_ratio=$KD_RATIO"

TUNE_CMD="tune run"

mkdir -p "$CHECKPOINT_OUTPUT_DIR"
touch "$CHECKPOINT_OUTPUT_DIR/.inprogress"

CMD="$TUNE_CMD $DIST_TRAIN_OPTIONS $TRAIN_METHOD $CONFIG_FILE_OPTION $CONFIG_OPTIONS"

```
</details> 

Thank you.

**Review by ebsmothers — APPROVED** [review]
Thanks @insop, this looks good to me! Once CI is green I think we're good to merge here

**Comment by ebsmothers** [comment]
Hi @insop do you mind pulling latest main and merging the changes into your PR? We had a breakage that should be fixed now, just want to make sure we can get green CI prior to merging this PR. Thanks!

**Comment by insop** [comment]
> Hi @insop do you mind pulling latest main and merging the changes into your PR? We had a breakage that should be fixed now, just want to make sure we can get green CI prior to merging this PR. Thanks!

Hi @ebsmothers,

Just merge to the upstream/master.
I will monitor the workflows.

Thank you,
